### ISSUE TITLE
Feature Request: specify custom chain spec at remote location

### ISSUE THREAD

**Issue by stone212** [paritytech/parity#7231]
A feature request. Currently you can specify a custom chain spec with:

$ parity --chain /path/to/chain-spec.json

The request is to allow:

$ parity --chain http://aws-for-example.com/chain-spec.json

This would allow multiple nodes to use one central chain spec file

**Comment by 5chdn** [comment]
Duplicate of #7210

**Comment by 5chdn** [comment]
This idea is not something we consider but pull requests are welcome.

**Comment by tomusdrw** [comment]
That works:
```
wget http://aws-for-example.com/chain-spec.json && parity --chain ./chain-spec.json
```

**Comment by stone212** [comment]
@tomusdrw No that doesn't work.  It accomplishes a different goal, which is to get the chain spec onto the localhost from a remote location.  My suggestion was to allow configuration from a remote file.  It's a significant difference.

@5chdn you said "pull requests are welcome" and normally that would make sense to me but I think I don't understand in this case.  Usually with git a person needs to have access to the repo to create a side branch and issue a pull request.  Are you saying that this repo is public?

**Comment by tomusdrw** [comment]
@stone212 the issue doesn't describe under what circumstances the file might change, Parity would need to poll and somehow update internal settings without interuption of regular work or you would need to restart.

Second, and most important - there is no good and generic enough way to ensure security of such setup. If http is used you are suspectible to mitm attack (the attacker will be able to connect your node to his bootnodes for instance), if you use https you rely on web2.0 security measures which might not be enough.

Solving this issues is beyond parity scope, you can simply create a script that does what you want in couple of lines of code with security guarantees you need. For instance the spec file hash could be stored on the blockchain and the file served over IPFS, whenever the file changes you fetch it and restart Parity.

> Are you saying that this repo is public?

Yes the repo is public, we accept external Pull Requests, feel free to fork the repo, commit to your fork and then propose changes as Pull Request.

**Comment by stone212** [comment]
> Second, and most important - there is no good and generic enough way to ensure security of such setup. 

If you think this means that it should not be included, then I accept the decision.  I think that it should be up to the user to make the choice but I understand the logic.  

> Yes the repo is public, we accept external Pull Requests, feel free to fork the repo, commit to your fork and then propose changes as Pull Request.

I see what you mean by "public" but I meant "public writeable" since we were talking about a commit.  But okay I looked up how GitHub handles pull requests.  It's more complicated than working in a small team but I guess it has to be.

**Comment by stone212** [comment]
@tomusdrw you said "the issue doesn't describe under what circumstances the file might change" and now I have a great example:

https://github.com/paritytech/parity/issues/7399#event-1403552991

If I start a private blockchain with 50 nodes now, then when I upgrade to Parity 1.9.0 I will have to copy the file 50 times.  If it was possible to invoke Parity to read a chain spec file remotely, then I could change it only once.

I still accept that you may not want to add the feature, but at least you see the reason for it.  If you need more resources, that is something I would like to discuss.

**Comment by tomusdrw** [comment]
@stone212 I still believe that the solution I provided:
```
wget http://aws-for-example.com/chain-spec.json && parity --chain ./chain-spec.json
```
works well in your example and doesn't lead to any security discussions within Parity. While the feature seems simple to implement the security tradeoffs associated with it are better left for the users.

Note that in my example I'm using http without any checksums so it's susceptible to man-in-the-middle attack (where attacker connects you to a completely different network and feeds you a malicious state snapshot), but it might be a tradeoff someone is willing to accept, while I would never support implementing something like this in Parity.

**Comment by stone212** [comment]
@tomusdrw I understand that you don't want to implement the feature but the wget example you provided is what I had already been using that led me to make this feature request.  Your example does nothing to address the actual request.  It is entirely unrelated.  I've said that before and I'm not sure why you think it addresses this issue.  Unless you're also suggesting that every parity node in a private blockchain has a cron job where on every boot it wget's the chain spec again, or something strange like that.

But I understand you don't want to implement the feature and I respect that.  I had been ready to assign a team to adding the feature ourselves to add to the Parity code as @5chdn suggested but from your reply I get the feeling that he was wrong to suggest that and you would not accept this change if we made it.  So I won't hire the team.

**Comment by kirushik** [comment]
_My 2¢:_

I think this would actually be an anti-feature, because it (even if slightly) replaces distributed control over what are the consensus rules with a centralized one.
What would happen if the centralized server is hacked? The attacker then would be able to force all the consumers of remote chainspec to mine `foundation` Ethereum to his coinbase instead of their expected function. Or introduce a malicious chain fork without any discussions with community.

In this perspective having to care about chainspec updates for node owners doesn't seem too bad, actually.
@stone212 I'll be glad to discuss some alternative solutions to your issue, something in line with what @tomusdrw proposed: storing the chainspec ipfs hash on blockchain and having on-chain mechanism for chainspec updates proposals and acceptance.

**Comment by stone212** [comment]
@kirushik I'd enjoy discussing possibilities.  I'm on gitter now.  I did not mean for this to balloon the way it has.  But I will also reply to what you wrote here:

> it (even if slightly) replaces distributed control over what are the consensus rules with a centralized one

And that would be a bad feature - for a public blockchain.  Not so much for a private one.
